src/Facebook/InstantArticles/Transformer/Rules/Rule.php

```diff
@@ -8,72 +8,87 @@
+        //     var_dump('context matches: '.($matches_context ? 'MATCHES' : 'no match'));
+        //     var_dump('node name: <'.$node->nodeName.' />');
+        //     var_dump('node: '.$node->ownerDocument->saveXML($node));
+        //     var_dump('node matches: '.($matches_node ? 'MATCHES' : 'no match'));
+        //     var_dump('rule: '.get_class($this));
+        //     var_dump('-------');
+        // }
         return $matches_context && $matches_node;
     }
 
-    abstract public function matchesContext($context);
+    abstract public function matchesContext(Element $context): bool;
 
-    abstract public function matchesNode($node);
+    abstract public function matchesNode(\DOMNode $node): bool;
 
-    abstract public function apply($transformer, $container, $node);
+    abstract public function apply(Transformer $transformer, Element $container, \DOMNode $node): Element;
 
-    abstract public function getContextClass();
+    abstract public function getContextClass(): Vector<string>;
 
-    public static function create()
+    public static function create(): Rule
     {
         throw new \Exception(
             'All Rule class extensions should implement the '.
             'Rule::create() method'
         );
     }
 
-    public static function createFrom($configuration)
+    public static function createFrom(array<string, mixed> $configuration): Rule
     {
         throw new \Exception(
             'All Rule class extensions should implement the '.
             'Rule::createFrom($configuration) method'
         );
     }
 
-    public static function retrieveProperty($array, $property_name)
+    public static function retrieveProperty(array<string, mixed> $properties, string $property_name): mixed
     {
-        if (isset($array[$property_name])) {
-            return $array[$property_name];
-        } elseif (isset($array['properties']) && isset($array['properties'][$property_name])) {
-            return $array['properties'][$property_name];
+        if (array_key_exists($property_name, $properties)) {
+            return $properties[$property_name];
+        } elseif (array_key_exists('properties', $properties)) {
+            $mappedProperties = $properties['properties'];
+            if (is_array($mappedProperties) && array_key_exists($property_name, $mappedProperties)) {
+                return $mappedProperties[$property_name];
+            }
         }
+        return null;
     }
 
     /**
      * Auxiliary method to extract full qualified class name.
      * @return string The full qualified name of class
      */
-    public static function getClassName()
+    public static function getClassName(): string
+    {
+        return get_called_class();
+    }
+
+    /**
+     * Auxiliary method to extract all Elements full qualified class name.
+     *
+     * @return string The full qualified name of class.
+     */
+    public function getObjClassName(): string
```

Comment:
How is that different from `getClassName()`? If it's a leftover, let's get rid of it.

Reply:
Unfortunately using //strict it forces to call each method properly.
so self::getClassName() and $instance->getObjClassName() are required for code =(